dataprep/eda/eda_frame.py

```diff
@@ -0,0 +1,368 @@
+"""Defines DataArray."""
+
+from functools import reduce
+from math import ceil
+from typing import Any, List, Optional, Sequence, Tuple, Union, cast, Dict
+from collections import Counter
+import warnings
+
+import dask
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+import pandas._libs.missing as libmissing
+
+from .dtypes_v2 import NUMERICAL_DTYPES, DType, DTypeDef, detect_dtype, Nominal, GeoGraphy
+
+DataFrame = Union[pd.DataFrame, dd.DataFrame, "EDAFrame"]
+
+
+class EDAFrame:
+    """EDAFrame provides an abstraction over dask DataFrame
+    and dask Array. The reason is that sometimes some algorithms
+    only works on the Array and not the DataFrame. However,
+    the cost for getting the array from a dask DataFrame (with known length)
+    is non trivial. Instead of computing the array from a dask
+    DataFrame again and again, it would be better do that once.
+
+    Other reasons to have a separate EDAFrame abstraction includes
+    converting the column names to string without modifying the
+    DataFrame from user, and preprocessings like dropna and type detection.
+
+    Parameters
+    ----------
+    df
+        The DataFrame
+    repartition
+        Whether to repartition the DataFrame into 128M chunks.
+    dtype: str or DType or dict of str or dict of DType, default None
+        Specify Data Types for designated column or all columns.
+        E.g.  dtype = {"a": Continuous, "b": "Nominal"} or
+        dtype = {"a": Continuous(), "b": "nominal"}
+        or dtype = Continuous() or dtype = "Continuous" or dtype = Continuous()
+    """
+
+    # pylint: disable = too-many-branches
+    # pylint: disable = too-many-statements
+    # pylint: disable = too-many-instance-attributes
+    def __init__(
+        self,
+        df: Optional[DataFrame] = None,
+        repartition: bool = True,
+        dtype: Optional[DTypeDef] = None,
+    ) -> None:
+
+        _suppress_warnings()
+
+        # Init. instance attribute.
+        self._ddf: Optional[dd.DataFrame] = None
+        self._values: Optional[da.Array] = None
+        self._nulls: Optional[Union[da.Array, np.ndarray]] = None
+        self._columns: Optional[pd.Index] = None
+        self._head: Optional[pd.DataFrame] = None
+        self._shape: Optional[Tuple[int, int]] = None
+        self._eda_dtypes: Dict[str, DType] = {}
+        self._str_col_cache: Dict[Tuple[str, bool], dd.Series] = {}
+        self._nulls_cnt: Dict[str, int] = {}
+
+        if df is None:
+            return
+
+        if isinstance(df, EDAFrame):
+            self._ddf = df._ddf
+            self._values = df._values
```

Comment:
It seems like we don't need values anymore? I searched the whole codebase and nowhere we touched this field.